packages/kit/src/core/dev/plugin.js

```diff
@@ -387,14 +387,36 @@ function remove_html_middlewares(server) {
 }
 
 /**
+ * @param {import('vite').ViteDevServer} vite
  * @param {import('vite').ModuleNode} node
  * @param {Set<import('vite').ModuleNode>} deps
  */
-function find_deps(node, deps) {
-	for (const dep of node.importedModules) {
-		if (!deps.has(dep)) {
-			deps.add(dep);
-			find_deps(dep, deps);
+function find_deps(vite, node, deps) {
+	// since `ssrTransformResult.deps` contains URLs instead of `ModuleNode`s, this process is asynchronous.
+	// instead of using `await`, we resolve all branches in parallel.
+	/** @type {Promise<void>[]} */
+	const branches = [];
+
+	/** @param {import('vite').ModuleNode} node */
+	function add(node) {
+		if (!deps.has(node)) {
+			deps.add(node);
+			branches.push(find_deps(vite, node, deps));
 		}
 	}
+
+	/** @param {string} url */
+	async function add_by_url(url) {
+		branches.push(vite.moduleGraph.getModuleByUrl(url).then((node) => node && add(node)));
+	}
+
+	if (node.ssrTransformResult) {
+		if (node.ssrTransformResult.deps) {
+			node.ssrTransformResult.deps.forEach(add_by_url);
+		}
+	} else {
+		node.importedModules.forEach(add);
+	}
+
+	return Promise.all(branches).then(() => {});
```

Comment:
This looks suspicious to me — it will only await the promises that are added to `branches` _before_ the `Promise.all`. Don't we need to await the ones added later?

Demo here — even though we push to the array before the first promise resolves, the result is `[1]` rather than `[1, 2]`:

```js
function resolve(value) {
  return Promise.resolve().then(() => {
    console.log(`resolving ${value}`);
    return value;
  });
}

array = [resolve(1)];

Promise.all(array).then((values) => {
  console.log('result', values);
});

array.push(resolve(2));

// resolving 1
// resolving 2
// result [1]
```

Given that we're mostly CPU-bound here, I wonder if we shouldn't just do everything serially?

Reply:
All pushes to `branches` happen synchronously—each `find_deps` call has their own `branches`.

I thought this would be IO bound, because otherwise `getModuleByUrl` wouldn't need to be async, but I'll gladly make this serial instead if you think it isn't worth it.

Reply:
d'oh! of course